### PR TITLE
Add user agent header to TS client and normalized the other clients

### LINKED ISSUE #12295: Add User-Agent Header to identify client types
We had this at some point and removed it, but I think it would be valuable to be able to tell what type of clients are connecting. We'd do this by setting the User-Agent header. This could definitely be a point of interest for customers/ASRS.

### PR DESCRIPTION
**Pull request by BrennanConroy**
Fixes https://github.com/aspnet/AspNetCore/issues/12295

One unfortunate thing is that it's difficult to suppress the user agent on the TS client right now because we don't expose any headers configuration.

Examples:
Browser: `Microsoft SignalR/5.0 (5.0.0-dev; Unknown OS; Browser; Unknown Runtime Version)`
Node: `Microsoft SignalR/5.0 (5.0.0-dev; Windows NT; NodeJS; 10.15.0)`
.NET: `Microsoft SignalR/42.42 (5.0.0-dev; Windows NT; .NET; .NET Core 5.0.0-alpha1.19470.7)`
Java: `Microsoft SignalR/99.99 (99.99.99-dev; Windows NT; Java; 11.0.3; JetBrains s.r.o)`

### CODE REVIEW

File: src/SignalR/clients/ts/signalr/src/Utils.ts [hunk 1]
```diff
@@ -181,3 +187,68 @@ export class ConsoleLogger implements ILogger {
         }
     }
 }
+
+/** @private */
+export function getUserAgentHeader(): [string, string] {
+    let userAgentHeaderName = "X-SignalR-UserAgent";
+    if (Platform.isNode) {
+        userAgentHeaderName = "User-Agent";
```
Comment (halter73) [inline src/SignalR/clients/ts/signalr/src/Utils.ts]: Why not just use "X-SignalR-UserAgent" on all platforms?

**Comment by halter73** [comment]
Do we have any tests verifying that the server sees the "X-SignalR-UserAgent" header on the negotiate/send/connect/poll requests from the various transports?

File: src/SignalR/clients/ts/FunctionalTests/ts/HubConnectionTests.ts [hunk 2]
```diff
@@ -1092,6 +1093,25 @@ describe("hubConnection", () => {
         }
     });
 
+    it("sets the user agent header", async (done) => {
+        const hubConnection = getConnectionBuilder(HttpTransportType.LongPolling, TESTHUB_NOWEBSOCKETS_ENDPOINT_URL)
```
Comment (halter73) [inline src/SignalR/clients/ts/FunctionalTests/ts/HubConnectionTests.ts]: Why use the `TESTHUB_NOWEBSOCKETS_ENDPOINT_URL` if WebSocketTransport.ts was changed to send the header?

File: src/SignalR/clients/ts/signalr/src/Utils.ts (hunk 1 above)
Comment (BrennanConroy) [reply src/SignalR/clients/ts/signalr/src/Utils.ts]: Because "X-SignalR-UserAgent" is a workaround for not being able to set the user-agent in browsers

File: src/SignalR/clients/ts/FunctionalTests/ts/HubConnectionTests.ts (hunk 2 above)
Comment (BrennanConroy) [reply src/SignalR/clients/ts/FunctionalTests/ts/HubConnectionTests.ts]: `<grumble>` because it requires some browser checks in the test `</grumble>`

I'll update it.

File: src/SignalR/clients/ts/signalr/src/Utils.ts (hunk 1 above)
Comment (anurse) [reply src/SignalR/clients/ts/signalr/src/Utils.ts]: Yeah, I'd rather use the real `User-Agent` header where we can.

**Comment by anurse** [comment]
Can we force every field up to runtime version to be filled in rather than just omitting blank fields? So for example, if `os` is unknown, instead of skipping it, write "Unknown OS". Ditto `runtime` ("Unknown Runtime") and `runtimeVersion` ("Unknown Version").

I'll put some comments in the test to explain even more clearly what I mean.

File: src/SignalR/clients/java/signalr/src/test/java/com/microsoft/signalr/UserAgentTest.java [hunk 3]
```diff
@@ -51,4 +51,17 @@ public void checkUserAgentString() {
...
+    @ParameterizedTest
+    @MethodSource("UserAgents")
+    public void UserAgentHeaderIsCorrect(String detailedVersion, String os, String runtime, String runtimeVersion, String vendor, String expected) {
+        assertEquals(expected, UserAgentHelper.constructUserAgentString(detailedVersion, os, runtime, runtimeVersion, vendor));
+    }
+
+    private static Stream<Arguments> UserAgents() {
+        return Stream.of(
+                Arguments.of("1.4.5-dev", "Windows NT", "Java", "7.0.1", "Oracle", "Microsoft SignalR/1.4 (1.4.5-dev; Windows NT; Java; 7.0.1; Oracle)"),
+                Arguments.of("3.1.0", "", "Java", "7.0.1", "", "Microsoft SignalR/3.1 (3.1.0; Java; 7.0.1)"),
```
Comment (anurse) [inline src/SignalR/clients/java/signalr/src/test/java/com/microsoft/signalr/UserAgentTest.java]:
```suggestion
                Arguments.of("3.1.0", "", "Java", "7.0.1", "", "Microsoft SignalR/3.1 (3.1.0; Unknown OS; Java; 7.0.1; Unknown Vendor)"),
```

File: src/SignalR/clients/ts/signalr/tests/UserAgent.test.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+import { constructUserAgent } from "../src/Utils";
+
+describe("User Agent", () => {
+    [["1.0.4-build.10", "Linux", "NodeJS", "10", "Microsoft SignalR/1.0 (1.0.4-build.10; Linux; NodeJS; 10)"],
+     ["1.4.7-build.10", "", "Browser", "", "Microsoft SignalR/1.4 (1.4.7-build.10; Browser)"],
```
Comment (anurse) [inline src/SignalR/clients/ts/signalr/tests/UserAgent.test.ts]:
```suggestion
     ["1.4.7-build.10", "", "Browser", "", "Microsoft SignalR/1.4 (1.4.7-build.10; Unknown OS; Browser; Unknown Version)"],
```

File: src/SignalR/clients/ts/signalr/tests/UserAgent.test.ts [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+     ["3.1.1-build.10", "macOS", "Browser", "", "Microsoft SignalR/3.1 (3.1.1-build.10; macOS; Browser)"],
```
Comment (anurse) [inline src/SignalR/clients/ts/signalr/tests/UserAgent.test.ts]:
```suggestion
     ["3.1.1-build.10", "macOS", "Browser", "", "Microsoft SignalR/3.1 (3.1.1-build.10; macOS; Browser; Unknown Version)"],
```

File: src/SignalR/clients/ts/signalr/tests/UserAgent.test.ts [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+     ["3.1.3-build.10", "", "Browser", "4", "Microsoft SignalR/3.1 (3.1.3-build.10; Browser; 4)"]]
```
Comment (anurse) [inline src/SignalR/clients/ts/signalr/tests/UserAgent.test.ts]:
```suggestion
     ["3.1.3-build.10", "", "Browser", "4", "Microsoft SignalR/3.1 (3.1.3-build.10; Unknown OS; Browser; 4)"]]
```

File: src/SignalR/server/SignalR/test/UserAgentHeaderTest.cs [hunk 7]
```diff
@@ -12,23 +14,18 @@ namespace Microsoft.AspNetCore.Http.Connections.Tests
...
-            var userAgent = Constants.UserAgentHeader;
-            var expectedUserAgent = $"Microsoft SignalR/{majorVersion}.{minorVersion} ({assemblyVersion.InformationalVersion}; {os}; {runtime}; {runtimeVersion})";
+            Assert.Equal(expected, Constants.ConstructUserAgent(version, detailedVersion, os, runtime, runtimeVersion));
+        }
 
-            Assert.Equal(expectedUserAgent, userAgent);
+        public static IEnumerable<object[]> UserAgents()
+        {
+            yield return new object[] { new Version(1, 4), "1.4.3-preview9", "Windows NT", ".NET", ".NET 4.8.7", "Microsoft SignalR/1.4 (1.4.3-preview9; Windows NT; .NET; .NET 4.8.7)" };
+            yield return new object[] { new Version(3, 1), "3.1.0", "", ".NET", ".NET 4.8.9", "Microsoft SignalR/3.1 (3.1.0; .NET; .NET 4.8.9)" };
```
Comment (anurse) [inline src/SignalR/server/SignalR/test/UserAgentHeaderTest.cs]:
```suggestion
            yield return new object[] { new Version(3, 1), "3.1.0", "", ".NET", ".NET 4.8.9", "Microsoft SignalR/3.1 (3.1.0; Unknown OS; .NET; .NET 4.8.9)" };
```

File: src/SignalR/server/SignalR/test/UserAgentHeaderTest.cs [hunk 8]
```diff
@@ -12,23 +14,18 @@ namespace Microsoft.AspNetCore.Http.Connections.Tests
...
+            yield return new object[] { new Version(3, 1), "3.1.0", "", ".NET", "", "Microsoft SignalR/3.1 (3.1.0; .NET)" };
```
Comment (anurse) [inline src/SignalR/server/SignalR/test/UserAgentHeaderTest.cs]:
```suggestion
            yield return new object[] { new Version(3, 1), "3.1.0", "", ".NET", "", "Microsoft SignalR/3.1 (3.1.0; Unknown OS; .NET; Unknown Version)" };
```

**Comment by anurse** [comment]
To expand even more on my thinking: I'm trying to make sure it's relatively deterministic to parse the User-Agent. When parsing up to the `runtime` element, we ensure everything looks the same. Then after the `runtime` element, we ensure things look the same **within each runtime**. For example, `Java` includes the vendor field but `.NET`, `NodeJS` and `Browser` don't. That's OK since once you get to `Java` in the `User-Agent`, the parsing code could adapt and change which fields it expects.

**Comment by BrennanConroy** [comment]
@aspnet-hello triage https://dev.azure.com/dnceng/public/_build/results?buildId=370180

**Comment by aspnet-hello** [comment]
This comment was made automatically. If there is a problem contact aspnetcore-build@microsoft.com.

I've triaged the above build. I've created/commented on the following issue(s)
https://github.com/aspnet/AspNetCore-Internal/issues/3172

File: src/SignalR/clients/ts/signalr/src/Utils.ts (hunk 1 above)
Comment (jkotalik) [reply src/SignalR/clients/ts/signalr/src/Utils.ts]: When do you get an error for setting the User-Agent?

Comment (jkotalik) [reply src/SignalR/clients/ts/signalr/src/Utils.ts]: Is there a way to try setting the User-Agent and if it fails, fallback to X-SignalR-UserAgent?

File: src/SignalR/clients/ts/signalr/src/Utils.ts [hunk 9]
```diff
@@ -181,3 +187,72 @@ export class ConsoleLogger implements ILogger {
         }
     }
 }
+
+/** @private */
+export function getUserAgentHeader(): [string, string] {
+    let userAgentHeaderName = "X-SignalR-UserAgent";
```
Comment (jkotalik) [inline src/SignalR/clients/ts/signalr/src/Utils.ts]: Why X-SignalR-UserAgent instead of X-SignalR-User-Agent to be more consistent with the original value?

File: src/SignalR/clients/ts/signalr/src/Utils.ts [hunk 10]
```diff
@@ -181,3 +187,72 @@ export class ConsoleLogger implements ILogger {
...
+    }
+
+    userAgent += ")";
+    return userAgent;
+}
+
+function getOsName(): string {
+    let os: string = "";
+    if (Platform.isNode) {
+        os = process.platform;
```
Comment (jkotalik) [inline src/SignalR/clients/ts/signalr/src/Utils.ts]: nit: can't you just put the switch inside of this if?

Comment (jkotalik) [reply src/SignalR/clients/ts/signalr/src/Utils.ts]: Also after a quick google search, there seem to be ways to know the os platform from pure javascript. Seems super hacky though: https://stackoverflow.com/questions/9514179/how-to-find-the-operating-system-version-using-javascript

File: src/SignalR/clients/ts/signalr/src/Utils.ts [hunk 11]
```diff
@@ -7,6 +7,10 @@ import { NullLogger } from "./Loggers";
 import { IStreamSubscriber, ISubscription } from "./Stream";
 import { Subject } from "./Subject";
 
+// Version token that will be replaced by the prepack command
+/** The version of the SignalR client. */
+export const VERSION: string = "0.0.0-DEV_BUILD";
```
Comment (jkotalik) [inline src/SignalR/clients/ts/signalr/src/Utils.ts]: Why was this moved from Index to Utils?

**Review by jkotalik — APPROVED** [review]

**Review by halter73 — APPROVED** [review]

File: src/SignalR/clients/ts/signalr/src/Utils.ts (hunk 1 above)
Comment (BrennanConroy) [reply src/SignalR/clients/ts/signalr/src/Utils.ts]: Looks like in more recent browser versions we should be able to set the User-Agent, it is no longer on the forbidden headers list: https://developer.mozilla.org/en-US/docs/Glossary/Forbidden_header_name

File: src/SignalR/clients/ts/signalr/src/Utils.ts (hunk 10 above)
Comment (BrennanConroy) [reply src/SignalR/clients/ts/signalr/src/Utils.ts]: Yeah, not doing that :)

File: src/SignalR/clients/ts/signalr/src/Utils.ts (hunk 11 above)
Comment (BrennanConroy) [reply src/SignalR/clients/ts/signalr/src/Utils.ts]: So that it could be referenced from other files. You can't really reference index.ts as it's exporting all the public types.

File: src/SignalR/clients/ts/signalr/src/Utils.ts (hunk 9 above)
Comment (BrennanConroy) [reply src/SignalR/clients/ts/signalr/src/Utils.ts]: I like that better :+1:

File: src/SignalR/clients/ts/signalr/src/Utils.ts (hunk 1 above)
Comment (BrennanConroy) [reply src/SignalR/clients/ts/signalr/src/Utils.ts]: I'm back to keeping X-SignalR-User-Agent, this way the browser can provide the OS and browser etc. in the User-Agent header and we can provide our own info in our header.
